File: crates/viben-core/src/gateway/routes/cron.rs
//! Cron job management endpoints

use axum::{
    Json, Router,
    extract::{Path, State},
    routing::{delete, get, post, patch},
};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::gateway::{AppState, GatewayError};
use crate::services::{CreateCronJob, CronJob, CronNotificationSettings, UpdateCronJob};

/// Cron job response
#[derive(Serialize)]
pub struct CronJobResponse {
    pub id: String,
    pub name: String,
    pub enabled: bool,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cron: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub every: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub channel: Option<String>,
    pub agent: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_run: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_run: Option<i64>,
    pub created_at: i64,
    pub updated_at: i64,
}

impl From<CronJob> for CronJobResponse {
    fn from(job: CronJob) -> Self {
        Self {
            id: job.id,
            name: job.name,
            enabled: job.enabled,
            message: job.message,
            cron: job.cron,
            every: job.every,
            channel: job.channel,
            agent: job.agent,
            last_run: job.last_run,
            last_status: job.last_status.map(|s| s.to_string()),
            last_error: job.last_error,
            next_run: job.next_run,
            created_at: job.created_at,
            updated_at: job.updated_at,
        }
    }
}

/// List cron jobs response
#[derive(Serialize)]
pub struct ListCronJobsResponse {
    pub jobs: Vec<CronJobResponse>,
}

/// List all cron jobs
pub async fn list_cron_jobs(
    State(state): State<AppState>,
) -> Result<Json<ListCronJobsResponse>, GatewayError> {
    tracing::debug!(target: "viben::gateway::cron", "Listing all cron jobs");

    let jobs = state.cron.list_jobs().await;
    let count = jobs.len();
    let job_responses: Vec<CronJobResponse> = jobs.into_iter().map(CronJobResponse::from).collect();

    tracing::debug!(target: "viben::gateway::cron", "Listed {} cron jobs", count);

    Ok(Json(ListCronJobsResponse {
        jobs: job_responses,
    }))
}

/// Get cron job by ID
pub async fn get_cron_job(
    State(state): State<AppState>,
    Path(job_id): Path<String>,
) -> Result<Json<CronJobResponse>, GatewayError> {
    tracing::debug!(target: "viben::gateway::cron", "Getting cron job: {}", job_id);

    let job = state.cron.get_job(&job_id).await.ok_or_else(|| {
        tracing::warn!(target: "viben::gateway::cron", "Cron job not found: {}", job_id);
        GatewayError::NotFound(format!("Cron job not found: {}", job_id))
    })?;

    tracing::trace!(
        target: "viben::gateway::cron",
        "Found cron job: {} (enabled={})",
        job_id, job.enabled
    );

    Ok(Json(CronJobResponse::from(job)))
}

/// Create cron job request
#[derive(Deserialize)]
pub struct CreateCronJobRequest {
    #[serde(default)]
    pub id: Option<String>,
    pub name: String,
    pub message: String,
    #[serde(default)]
    pub script: Option<String>,
    #[serde(default)]
    pub cron: Option<String>,
    #[serde(default)]
    pub every: Option<u64>,
    #[serde(default)]
    pub channel: Option<String>,
    #[serde(default)]
    pub agent: Option<String>,
    #[serde(default = "default_enabled")]
    pub enabled: bool,
    #[serde(default)]
    pub notifications: Option<CronNotificationSettings>,
}

fn default_enabled() -> bool {
    true
}

/// Create a new cron job
pub async fn create_cron_job(
    State(state): State<AppState>,
    Json(req): Json<CreateCronJobRequest>,
) -> Result<Json<CronJobResponse>, GatewayError> {
    tracing::info!(
        target: "viben::gateway::cron",
        "Creating new cron job: name='{}', cron={:?}, every={:?}",
        req.name, req.cron, req.every
    );

    let create_data = CreateCronJob {
        id: req.id,
        name: req.name.clone(),
        message: req.message,
        script: req.script,
        cron: req.cron,
        every: req.every,
        channel: req.channel,
        agent: req.agent,
        enabled: req.enabled,
        notifications: req.notifications,
    };

    let job = state.cron.create_job(create_data).await?;

    tracing::info!(
        target: "viben::gateway::cron",
        "Cron job created: id={}, name='{}'",
        job.id, job.name
    );

    Ok(Json(CronJobResponse::from(job)))
}

/// Update cron job request
#[derive(Deserialize)]
pub struct UpdateCronJobRequest {
    pub name: Option<String>,
    pub message: Option<String>,
    pub script: Option<String>,
    pub cron: Option<String>,
    pub every: Option<u64>,
    pub channel: Option<String>,
    pub agent: Option<String>,
    pub enabled: Option<bool>,
    pub notifications: Option<CronNotificationSettings>,
}

/// Update an existing cron job
pub async fn update_cron_job(
    State(state): State<AppState>,
    Path(job_id): Path<String>,
    Json(req): Json<UpdateCronJobRequest>,
) -> Result<Json<CronJobResponse>, GatewayError> {
    tracing::debug!(
        target: "viben::gateway::cron",
        "Updating cron job: {} (name={:?}, enabled={:?})",
        job_id, req.name, req.enabled
    );

    let update_data = UpdateCronJob {
        name: req.name,
        message: req.message,
        script: req.script,
        cron: req.cron,
        every: req.every,
        channel: req.channel,
        agent: req.agent,
        enabled: req.enabled,
        notifications: req.notifications,
    };

    let job = state.cron.update_job(&job_id, update_data).await?;

    tracing::debug!(target: "viben::gateway::cron", "Cron job {} updated successfully", job_id);

    Ok(Json(CronJobResponse::from(job)))
}

/// Delete a cron job
pub async fn delete_cron_job(
    State(state): State<AppState>,
    Path(job_id): Path<String>,
) -> Result<Json<Value>, GatewayError> {
    tracing::info!(target: "viben::gateway::cron", "Deleting cron job: {}", job_id);

    state.cron.delete_job(&job_id).await?;

    tracing::info!(target: "viben::gateway::cron", "Cron job {} deleted", job_id);

    Ok(Json(json!({
        "deleted": job_id
    })))
}

/// Enable a cron job
pub async fn enable_cron_job(
    State(state): State<AppState>,
    Path(job_id): Path<String>,
) -> Result<Json<CronJobResponse>, GatewayError> {
    tracing::info!(target: "viben::gateway::cron", "Enabling cron job: {}", job_id);

    let job = state.cron.enable_job(&job_id).await?;

    tracing::info!(target: "viben::gateway::cron", "Cron job {} enabled", job_id);

    Ok(Json(CronJobResponse::from(job)))
}

/// Disable a cron job
pub async fn disable_cron_job(
    State(state): State<AppState>,
    Path(job_id): Path<String>,
) -> Result<Json<CronJobResponse>, GatewayError> {
    tracing::info!(target: "viben::gateway::cron", "Disabling cron job: {}", job_id);

    let job = state.cron.disable_job(&job_id).await?;

    tracing::info!(target: "viben::gateway::cron", "Cron job {} disabled", job_id);

    Ok(Json(CronJobResponse::from(job)))
}

/// Run a cron job immediately
pub async fn run_cron_job(
    State(state): State<AppState>,
    Path(job_id): Path<String>,
) -> Result<Json<Value>, GatewayError> {
    tracing::info!(target: "viben::gateway::cron", "Running cron job immediately: {}", job_id);

    state.cron.run_job(&job_id).await?;

    tracing::info!(target: "viben::gateway::cron", "Cron job {} execution triggered", job_id);

    Ok(Json(json!({
        "triggered": job_id,
        "message": "Job execution started"
    })))
}

/// Create the cron router
pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/cron", get(list_cron_jobs))
        .route("/api/cron", post(create_cron_job))
        .route("/api/cron/:id", get(get_cron_job))
        .route("/api/cron/:id", patch(update_cron_job))
        .route("/api/cron/:id", delete(delete_cron_job))
        .route("/api/cron/:id/enable", post(enable_cron_job))
        .route("/api/cron/:id/disable", post(disable_cron_job))
        .route("/api/cron/:id/run", post(run_cron_job))
}

File: packages/core/src/gateway/routes/cron.ts
/**
 * Cron routes
 */
import type { FastifyInstance } from "fastify";
import type { CronJob, CreateCronJob, UpdateCronJob } from "../../services/cron";
import type { AppState } from "../state";
import { trace, SpanStatusCode } from "../../telemetry";
import { getSpanName } from "../../telemetry/route-names";

// Get tracer for cron routes
const tracer = trace.getTracer("viben-gateway", "1.0.0");

/**
 * Transform CronJob to snake_case response format (to match Rust gateway)
 * Note: CronJob interface now uses snake_case, so this is a simple passthrough
 */
function toSnakeCaseJob(job: CronJob) {
  return {
    id: job.id,
    name: job.name,
    enabled: job.enabled,
    job_type: job.job_type,
    message: job.message,
    script: job.script,
    cron: job.cron,
    every: job.every,
    channel: job.channel,
    agent: job.agent,
    workspace_path: job.workspace_path,
    notifications: job.notifications,
    last_run: job.last_run,
    last_status: job.last_status,
    last_error: job.last_error,
    last_output: job.last_output,
    next_run: job.next_run,
    created_at: job.created_at,
    updated_at: job.updated_at,
  };
}

/**
 * Register cron routes
 */
export function registerCronRoutes(fastify: FastifyInstance, state: AppState): void {
  // List all cron jobs
  fastify.get("/api/cron", {
    schema: {
      description: "List all cron jobs",
      tags: ["cron"],
      response: {
        200: {
          type: "object",
          properties: {
            jobs: {
              type: "array",
              items: {
                type: "object",
                properties: {
                  id: { type: "string" },
                  name: { type: "string" },
                  enabled: { type: "boolean" },
                  job_type: { type: "string", enum: ["agent", "script", "message"] },
                  message: { type: "string" },
                  script: { type: "string" },
                  cron: { type: "string", description: "Cron expression" },
                  every: { type: "number", description: "Interval in milliseconds" },
                  channel: { type: "string" },
                  agent: { type: "string" },
                  workspace_path: { type: "string" },
                  notifications: { type: "object" },
                  last_run: { type: "number" },
                  last_status: { type: "string" },
                  last_error: { type: "string" },
                  last_output: { type: "string" },
                  next_run: { type: "number" },
                  created_at: { type: "number" },
                  updated_at: { type: "number" },
                },
              },
            },
          },
        },
      },
    },
  }, async () => {
    const span = tracer.startSpan(getSpanName("cron.list"));
    try {
      const jobs = await state.cron.listJobs();
      span.setAttributes({
        "cron.job_count": jobs.length,
        "cron.enabled_count": jobs.filter(j => j.enabled).length,
      });
      span.setStatus({ code: SpanStatusCode.OK });
      return { jobs: jobs.map(toSnakeCaseJob) };
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to list cron jobs" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      throw e;
    } finally {
      span.end();
    }
  });

  // Get a specific cron job
  fastify.get<{ Params: { id: string } }>("/api/cron/:id", {
    schema: {
      description: "Get a specific cron job by ID",
      tags: ["cron"],
      params: {
        type: "object",
        properties: {
          id: { type: "string", description: "Cron job ID" },
        },
        required: ["id"],
      },
      response: {
        200: {
          type: "object",
          properties: {
            id: { type: "string" },
            name: { type: "string" },
            enabled: { type: "boolean" },
            job_type: { type: "string" },
            message: { type: "string" },
            script: { type: "string" },
            cron: { type: "string" },
            every: { type: "number" },
            channel: { type: "string" },
            agent: { type: "string" },
            workspace_path: { type: "string" },
            notifications: { type: "object" },
            last_run: { type: "number" },
            last_status: { type: "string" },
            last_error: { type: "string" },
            last_output: { type: "string" },
            next_run: { type: "number" },
            created_at: { type: "number" },
            updated_at: { type: "number" },
          },
        },
        404: {
          type: "object",
          properties: {
            error: { type: "string" },
          },
        },
      },
    },
  }, async (request, reply) => {
    const { id } = request.params;
    const span = tracer.startSpan(getSpanName("cron.get"), {
      attributes: { "cron.job_id": id },
    });
    try {
      const job = await state.cron.getJob(id);
      if (!job) {
        span.setAttributes({ "cron.job_found": false });
        span.setStatus({ code: SpanStatusCode.ERROR, message: "Cron job not found" });
        reply.code(404);
        return { error: `Cron job not found: ${id}` };
      }
      span.setAttributes({
        "cron.job_found": true,
        "cron.job_name": job.name,
        "cron.job_type": job.job_type,
        "cron.job_enabled": job.enabled,
        "cron.job_agent": job.agent,
      });
      span.setStatus({ code: SpanStatusCode.OK });
      return toSnakeCaseJob(job);
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to get cron job" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      throw e;
    } finally {
      span.end();
    }
  });

  // Create a new cron job
  fastify.post<{ Body: CreateCronJob }>("/api/cron", async (request, reply) => {
    const input = request.body;
    const span = tracer.startSpan(getSpanName("cron.create"), {
      attributes: {
        "cron.job_name": input.name,
        "cron.job_type": input.job_type || "agent",
        "cron.job_enabled": input.enabled !== false,
        "cron.job_cron": input.cron || "",
        "cron.job_every": input.every || 0,
        "cron.job_agent": input.agent || "main",
      },
    });
    try {
      const job = await state.cron.createJob(input);
      span.setAttributes({
        "cron.job_id": job.id,
        "cron.job_created": true,
      });
      span.setStatus({ code: SpanStatusCode.OK });
      reply.code(201);
      return toSnakeCaseJob(job);
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to create cron job" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to create cron job" };
    } finally {
      span.end();
    }
  });

  // Update a cron job
  fastify.patch<{ Params: { id: string }; Body: UpdateCronJob }>("/api/cron/:id", async (request, reply) => {
    const { id } = request.params;
    const updates = request.body;
    const span = tracer.startSpan(getSpanName("cron.update"), {
      attributes: {
        "cron.job_id": id,
        "cron.update_fields": Object.keys(updates).join(","),
      },
    });
    try {
      const job = await state.cron.updateJob(id, updates);
      span.setAttributes({
        "cron.job_name": job.name,
        "cron.job_type": job.job_type,
        "cron.job_enabled": job.enabled,
      });
      span.setStatus({ code: SpanStatusCode.OK });
      return toSnakeCaseJob(job);
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to update cron job" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to update cron job" };
    } finally {
      span.end();
    }
  });

  // Delete a cron job
  fastify.delete<{ Params: { id: string } }>("/api/cron/:id", async (request, reply) => {
    const { id } = request.params;
    const span = tracer.startSpan(getSpanName("cron.delete"), {
      attributes: { "cron.job_id": id },
    });
    try {
      await state.cron.deleteJob(id);
      span.setStatus({ code: SpanStatusCode.OK });
      return { deleted: id };
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to delete cron job" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to delete cron job" };
    } finally {
      span.end();
    }
  });

  // Enable a cron job
  fastify.post<{ Params: { id: string } }>("/api/cron/:id/enable", async (request, reply) => {
    const { id } = request.params;
    const span = tracer.startSpan(getSpanName("cron.enable"), {
      attributes: { "cron.job_id": id },
    });
    try {
      const job = await state.cron.enableJob(id);
      span.setAttributes({
        "cron.job_name": job.name,
        "cron.job_enabled": job.enabled,
        "cron.job_next_run": job.next_run || 0,
      });
      span.setStatus({ code: SpanStatusCode.OK });
      return toSnakeCaseJob(job);
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to enable cron job" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to enable cron job" };
    } finally {
      span.end();
    }
  });

  // Disable a cron job
  fastify.post<{ Params: { id: string } }>("/api/cron/:id/disable", async (request, reply) => {
    const { id } = request.params;
    const span = tracer.startSpan(getSpanName("cron.disable"), {
      attributes: { "cron.job_id": id },
    });
    try {
      const job = await state.cron.disableJob(id);
      span.setAttributes({
        "cron.job_name": job.name,
        "cron.job_enabled": job.enabled,
      });
      span.setStatus({ code: SpanStatusCode.OK });
      return toSnakeCaseJob(job);
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to disable cron job" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to disable cron job" };
    } finally {
      span.end();
    }
  });

  // Run a cron job immediately
  fastify.post<{ Params: { id: string } }>("/api/cron/:id/run", async (request, reply) => {
    const { id } = request.params;
    const span = tracer.startSpan(getSpanName("cron.run"), {
      attributes: { "cron.job_id": id },
    });
    try {
      // Get job info before running for logging
      const job = await state.cron.getJob(id);
      if (job) {
        span.setAttributes({
          "cron.job_name": job.name,
          "cron.job_type": job.job_type,
          "cron.job_agent": job.agent,
        });
      }
      await state.cron.runJob(id);
      span.setStatus({ code: SpanStatusCode.OK });
      return { triggered: id, message: "Job execution started" };
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to run cron job" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to run cron job" };
    } finally {
      span.end();
    }
  });

  // Get execution logs for a cron job
  fastify.get<{
    Params: { id: string };
    Querystring: { limit?: string; offset?: string };
  }>("/api/cron/:id/logs", async (request, reply) => {
    const { id } = request.params;
    const limit = request.query.limit ? parseInt(request.query.limit, 10) : 100;
    const offset = request.query.offset ? parseInt(request.query.offset, 10) : 0;
    const span = tracer.startSpan(getSpanName("cron.logs"), {
      attributes: {
        "cron.job_id": id,
        "cron.logs_limit": limit,
        "cron.logs_offset": offset,
      },
    });
    try {
      const logs = await state.cron.getExecutionLogs(id, limit, offset);
      span.setAttributes({
        "cron.logs_count": logs.length,
      });
      span.setStatus({ code: SpanStatusCode.OK });
      return { logs };
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to get execution logs" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to get execution logs" };
    } finally {
      span.end();
    }
  });

  // Clear execution logs for a cron job
  fastify.delete<{ Params: { id: string } }>("/api/cron/:id/logs", async (request, reply) => {
    const { id } = request.params;
    const span = tracer.startSpan(getSpanName("cron.logs_clear"), {
      attributes: { "cron.job_id": id },
    });
    try {
      await state.cron.clearExecutionLogs(id);
      span.setStatus({ code: SpanStatusCode.OK });
      return { cleared: id };
    } catch (e) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: e instanceof Error ? e.message : "Failed to clear execution logs" });
      span.recordException(e instanceof Error ? e : new Error(String(e)));
      reply.code(400);
      return { error: e instanceof Error ? e.message : "Failed to clear execution logs" };
    } finally {
      span.end();
    }
  });
}
